---
title: 
output: html_document
---
 
```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(dplyr)
library(kableExtra)
library(DT)
library(downloadthis)
library(ggplot2)
```

The ATAC libraries are sequenced on multiple flow cells each. This may be done to increase sequencing depth, when the first sequencing run did not produce enough raw read pairs per cell. Here we would run cellranger-atac mkfastq a total of multiple times: once for each of the ATAC flow cells. All of the reads can be combined in a single instance of the cellranger-atac count pipeline. 

In the grouped pipeline, one sample is processed through one GEM well, resulting in one library which is sequenced across multiple flowcells. This workflow is commonly performed to increase sequencing depth. Finally, the pipelines perform secondary analyses like:

- Normalization of input runs to same median fragments per cell (sensitivity)
- Detection of accessible chromatin peaks
- Count matrix generation for peaks and transcription factors for the aggregate data
- Dimensionality reduction
- Cell clustering
- Cluster differential accessibility

```{r, echo = FALSE}
appliation='Cell Ranger ATAC count v2.0';

DiagrammeR::grViz("
  digraph graph2 {
  
  graph [layout = dot, rankdir = LR]
  
  # node definitions with substituted label text
  node [shape = box, style=rounded, color=brown, fillcolor=azure3, fontsize=25]
  a [label = '@@1']
  b [label = '@@2']
  c [label = '@@3']
  d [label = '@@4']
  e [label = '@@5']
  
  a -> b -> c -> d -> e
  }
  
  [1]: 'cellranger-atac mkfastq'
  [2]: appliation[1]
  [3]: 'Generation of metadata'
  [4]: 'Combining multiple sequencing runs'
  [5]: 'Secondary analyses'
  ",
  height = 100)
```


