---
title: 
output: html_document
---
 
```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(dplyr)
library(kableExtra)
library(DT)
library(downloadthis)
library(ggplot2)
```

Chromium Single Cell ATAC (Assay for Transposase Accessible Chromatin) allows you to analyze chromatin accessibility at the single cell level, providing insights into cell types and states, and deeper understanding of gene regulatory mechanisms. 

Cell Ranger ATAC analysis pipeline performs identification of open chromatin regions, motif annotation, and differential accessibility analysis for Single Cell ATAC data.

Cellranger-atac count takes FASTQ files from cellranger-atac mkfastq and performs ATAC analysis, including:

- Read filtering and alignment
- Barcode counting
- Identification of transposase cut sites
- Detection of accessible chromatin peaks
- Cell calling
- Count matrix generation for peaks and transcription factors
- Dimensionality reduction
- Cell clustering
- Cluster differential accessibility

```{r, echo = FALSE}
appliation='Cell Ranger ATAC count v2.0';

DiagrammeR::grViz("
  digraph graph2 {
  
  graph [layout = dot, rankdir = LR]
  
  # node definitions with substituted label text
  node [shape = box, style=rounded, color=brown, fillcolor=azure3, fontsize=25]
  a [label = '@@1']
  b [label = '@@2']
  c [label = '@@3']
  d [label = '@@4']
  
  a -> b -> c -> d
  }
  
  [1]: 'cellranger-atac mkfastq'
  [2]: appliation[1]
  [3]: 'Generation of metadata'
  [4]: 'ATAC analysis'
  ",
  height = 100)
```


