---
title: 
output: html_document
---
 
```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(dplyr)
library(kableExtra)
library(DT)
library(downloadthis)
library(ggplot2)
```

Cellranger count takes FASTQ files from cellranger mkfastq and performs alignment, filtering, barcode counting, and UMI counting. It uses the Chromium cellular barcodes to generate feature-barcode matrices, determine clusters, and perform gene expression analysis. In the count pipeline, one sample is processed through one GEM well and sequenced on one flowcell. In this case, generate FASTQs using cellranger mkfastq and run cellranger count as Single-Sample Analysis. Additionally, the pipelines perform secondary analyses like dimensionality reduction, clustering, and differential analysis.

```{r, echo = FALSE}
appliation='CellRanger/7.0 count';

DiagrammeR::grViz("
  digraph graph2 {
  
  graph [layout = dot, rankdir = LR]
  
  # node definitions with substituted label text
  node [shape = box, style=rounded, color=brown, fillcolor=azure3, fontsize=25]
  a [label = '@@1']
  b [label = '@@2']
  c [label = '@@3']
  d [label = '@@4']
  
  a -> b -> c -> d
  }
  
  [1]: 'bcl2fastq'
  [2]: appliation[1]
  [3]: 'Generation of metadata'
  [4]: 'Secondary analyses'
  ",
  height = 100)
```


