---
title: 
output: html_document
---
 
```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(dplyr)
library(kableExtra)
library(DT)
library(downloadthis)
library(ggplot2)
```

In the grouped count pipeline, one sample is processed through one GEM well, resulting in one library which is sequenced across multiple flowcells. This workflow is commonly performed to increase sequencing depth. In this case, all reads can be combined in a single instance of the cellranger count pipeline. Finally, the pipelines perform secondary analyses like dimensionality reduction, clustering, and differential analysis.

```{r, echo = FALSE}
appliation='CellRanger/7.0 count';

DiagrammeR::grViz("
  digraph graph2 {
  
  graph [layout = dot, rankdir = LR]
  
  # node definitions with substituted label text
  node [shape = box, style=rounded, color=brown, fillcolor=azure3, fontsize=25]
  a [label = '@@1']
  b [label = '@@2']
  c [label = '@@3']
  d [label = '@@4']
  e [label = '@@5']
  
  a -> b -> c -> d -> e
  }
  
  [1]: 'bcl2fastq'
  [2]: appliation[1]
  [3]: 'Generation of metadata'
  [4]: 'Merging multiple sequencing runs'
  [5]: 'Secondary analyses'
  ",
  height = 100)
```


